### PR TITLE
fix(api): PAPIv2.14: check for labware and module placement conflicts

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

Makes PAPIv2.14 protocols raise an error when they try to load modules or labware in ways that conflict with preexisting modules or labware. This matches existing behavior from PAPIv2.13 and below. Closes RCORE-531.

# Changelog

* **Protocol Engine core:** When loading labware and modules, check for conflicts with other labware and modules. This maps to the same underlying module as the legacy core for the business logic.
* **Legacy core:** a few user-facing errors were `AssertionError`s. Refactor them to `RuntimeError`s to make it clearer that they *will* be hit in practice.
* **LabwareView:** Refactor `get_definition_uri()` to return a `LabwareUri` instead of a plain `str`.

# Test plan

Here's a basic test to confirm that PAPIv2.14 now disallows things that would cause deck conflicts:

```python
requirements = {"apiLevel": "2.14"}

def run(context):
    # Should be OK:
    labware_1 = context.load_labware("opentrons_96_tiprack_300ul", 1)
    labware_2 = context.load_labware("opentrons_96_tiprack_300ul", 2)

    # Uncommenting should cause an error. Same deck slot as labware_2.
    # labware_3 = context.load_labware("opentrons_96_tiprack_300ul", 2)

    # Should be OK:
    heater_shaker = context.load_module("heaterShakerModuleV1", 3)

    # Uncommenting should cause an error. Open deck slot, but too close to heater_shaker.
    # magnetic_module = context.load_module("magnetic module gen2", 6)
```

I've been manually uploading this to a dev server, so far.

Separately, we need to make sure that when Labware Position Check loads labware into the run, the PAPIv2.14 protocol will *ignore* those labware for the purposes of deck conflict checking. I've included a Tavern integration test for this.

# Review requests

* As discussed in Slack, there's a fundamental problem with this implementation. By the time it has enough information to check for a deck conflict, the item has already been loaded, and the conflict is too late to prevent. I've tried to document this in the code with `# todo` comments. Do those comments make sense?
* Other than the mess described above, do the changes in `ProtocolCore` make sense? Did I miss any spots? The changes in that file are the least-well-tested part of this PR.

# Risk assessment

There's low risk that this will break anything that was working before in prior Python Protocol API versions.

But, there's a solid chance that this addition of conflict checking to PAPIv2.14 isn't quite right.

To resolve that risk, we probably need a way to automate the manual test protocol that I listed above. We also probably need to extend it to cover more cases. **Because of time constraints, that's not done in this PR.**

### CODE REVIEW

File: api/src/opentrons/protocol_api/core/engine/deck_conflict.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
+"""A Protocol-Engine-friendly wrapper for opentrons.motion_planning.deck_conflict."""
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_api/core/engine/deck_conflict.py:1]: ~~Reminder to self: disable these checks for OT-3, per https://github.com/Opentrons/opentrons/pull/12169#pullrequestreview-1311639861.~~

Comment (SyntaxColoring) [reply api/src/opentrons/protocol_api/core/engine/deck_conflict.py:1]: ~~Another reminder to self: add tests for protocol core.~~

File: api/src/opentrons/protocol_api/core/engine/deck_conflict.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+    new_module_id: Optional[str] = None,
+) -> None:
+    """Check for conflicts between items on the deck.
+
+    This is a Protocol-Engine-friendly wrapper around
+    opentrons.motion_planning.deck_conflict.check().
+
+    Params:
+        engine_state: An interface to retrieve details about the deck items.
+        robot_type: The type of robot these items are being placed in. Different
```
Comment (jbleon95) [inline api/src/opentrons/protocol_api/core/engine/deck_conflict.py]: `robot_type` no longer a param

Comment (SyntaxColoring) [reply api/src/opentrons/protocol_api/core/engine/deck_conflict.py]: TY!

**Review by mcous — APPROVED** [review]
Code changes LGTM. Good to merge from my end pending on-hardware testing

**Review by jbleon95 — APPROVED** [review]
Looks good to my eyes. I think the noted jankiness/known issues are contained enough that we can punt that for future improvements and changes we will make to this.

**Review by TamarZanzouri — APPROVED** [review]
Nice work! looks good to me!

**Comment by SyntaxColoring** [comment]
Merging without testing on hardware because something seems to have broken in my SSH configuration? Talking about this in-person, we think dev server testing is probably fine, anyway.

**Pull request merged**